### ISSUE TITLE
DM Multipath devices incorrectly considered as empty and available for OSD creation

### ISSUE THREAD

**Issue by dahorak** [ceph/ceph-ansible#1403]
_OSD auto discovery_ mode doesn't work on machines with DM Multipath devices.

I would like to use ceph-ansible for installation and configuration of Ceph cluster in [CentOS CI](https://wiki.centos.org/QaWiki/CI/) infrastructure. But it fails when `osd_auto_discovery` is set to `true`, because the machines used as OSD nodes have two disks (`sda` and `sdb`) used as DM Multipath device and ceph-ansible incorrectly consider them as empty/available.

This is the failed task:
```
...
TASK [ceph-osd : automatic prepare filestore osd disk(s) without partitions with collocated osd data and journal] ***
task path: /home/usmqe/ceph-ansible/roles/ceph-osd/tasks/scenarios/journal_collocation.yml:12
...
failed: [n49.crusty] (item={'key': u'sdb', 'value': {u'scheduler_mode': u'cfq', u'sectorsize': u'512', u'vendor': u'ATA', u'sectors': u'937426944', u'sas_device_handle': None, u'sas_address': None, u'host': u'SATA controller: Intel Corporation 82801IB (ICH9) 4 port SATA Controller [AHCI mode]', u'rotational': u'0', u'removable': u'0', u'support_discard': u'0', u'holders': [u'mpatha'], u'size': u'447.00 GB', u'model': u'SMvD1343095633FE', u'partitions': {}}}) => {
    "changed": true, 
    "cmd": [
        "ceph-disk", 
        "prepare", 
        "--cluster", 
        "test_name", 
        "/dev/sdb"
    ], 
    "delta": "0:00:00.465563", 
    "end": "2017-03-29 09:36:17.399985", 
    "failed": true, 
    "invocation": {
        "module_args": {
            "_raw_params": "ceph-disk prepare --cluster \"test_name\" \"/dev/sdb\"", 
            "_uses_shell": false, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "warn": true
        }, 
        "module_name": "command"
    }, 
    "item": {
        "key": "sdb", 
        "value": {
            "holders": [
                "mpatha"
            ], 
            "host": "SATA controller: Intel Corporation 82801IB (ICH9) 4 port SATA Controller [AHCI mode]", 
            "model": "SMvD1343095633FE", 
            "partitions": {}, 
            "removable": "0", 
            "rotational": "0", 
            "sas_address": null, 
            "sas_device_handle": null, 
            "scheduler_mode": "cfq", 
            "sectors": "937426944", 
            "sectorsize": "512", 
            "size": "447.00 GB", 
            "support_discard": "0", 
            "vendor": "ATA"
        }
    }, 
    "rc": 1, 
    "start": "2017-03-29 09:36:16.934422", 
    "stderr": "ceph-disk: Error: Device /dev/sdb is in use by a device-mapper mapping (dm-crypt?): dm-0", 
    "stdout": "", 
    "stdout_lines": [], 
    "warnings": []
}
...
```

And this is the partition schema and DM Multipath configuration from the node:
```
# lsblk /dev/sda /dev/sdb
NAME              MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda                 8:0    0   447G  0 disk  
└─mpatha          253:0    0   447G  0 mpath 
  ├─mpatha1       253:1    0   500M  0 part  /boot
  └─mpatha2       253:2    0 446.5G  0 part  
    ├─vg_n49-root 253:3    0 444.4G  0 lvm   /
    └─vg_n49-swap 253:4    0   2.1G  0 lvm   [SWAP]
sdb                 8:16   0   447G  0 disk  
└─mpatha          253:0    0   447G  0 mpath 
  ├─mpatha1       253:1    0   500M  0 part  /boot
  └─mpatha2       253:2    0 446.5G  0 part  
    ├─vg_n49-root 253:3    0 444.4G  0 lvm   /
    └─vg_n49-swap 253:4    0   2.1G  0 lvm   [SWAP]

# multipath -ll
mpatha (SMvD1343095633FE_1343095633FE) dm-0 ATA     ,SMvD1343095633FE
size=447G features='0' hwhandler='0' wp=rw
`-+- policy='service-time 0' prio=1 status=active
  |- 0:0:0:0  sda 8:0  active ready running
  `- 16:0:0:0 sdb 8:16 active ready running
```

**The problem is**, that DM Multipath partitions are not listed in `ansible_facts.ansible_devices[sda].partitions` (as it is also visible above in the failed task output: `"partitions": {},`), so following condition pass and the device is considered as empty (in `check_devices_auto.yml` and on other places).
```
  when:
    - item.value.partitions|count == 0
```

Possible solution might be to check also the list of _holders_ `ansible_facts.ansible_devices[sda].holders`: ["mpatha"],`).

----------------
_NOTE:_ On each of the OSD machine are empty devices `sdc`...`sdg` designated for OSDs (created virtually as iSCSI devices)
```
# lsblk /dev/sd{c..g}
NAME              MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sdc                 8:32   0    50G  0 disk  
sdd                 8:48   0    50G  0 disk  
sde                 8:64   0    50G  0 disk  
sdf                 8:80   0    50G  0 disk  
sdg                 8:96   0    50G  0 disk  
```

**Comment by mbukatov** [comment]
Looking at [source code of get_holders function](https://github.com/ansible/ansible/blob/097173c6f5e7dfebf445b6ac08f59eb1ea1df65f/lib/ansible/module_utils/facts.py#L1413) (which is necessary because the full list of facts gathered during [setup](https://docs.ansible.com/ansible/setup_module.html) [is not documented anywhere](https://docs.ansible.com/ansible/playbooks_variables.html)), it seems that `holders` value in `ansible_facts.ansible_devices[$device].holders` is based on [sysfs](https://en.wikipedia.org/wiki/Sysfs) `/sys/block/$device/holders` directory, which represent [which devices depend on which others](https://www.redhat.com/archives/dm-devel/2006-February/msg00091.html).

In other words, reporting device with non empty holders value as empty or available to use is just wrong.

**Comment by leseb** [comment]
I see two issues here:

* one with ceph-disk
* one with ansible

ceph-ansible uses both, but these problems are not ceph-ansible's fault, this is just the result. Can we have separate issues being opened on the respective projects to address this?

And btw, using autodiscovery or not, it's wrong to expect that ceph-disk will create a new lv on that device and bootstrap an OSD. ceph-disk doesn't play well with lvm. It expects either a disk or a directory. But I agree the issue is due to a wrong detection from Ansible.

Let me send a patch to use the holders as suggested.

**Comment by leseb** [comment]
Forget about what I said on ceph-disk, the error is expected, nothing to report, although I believe it's worth having an Ansible bug.
